### PR TITLE
Fnty args rustdoc

### LINKED ISSUE #44570: Callback parameter names are missing from rustdoc
Reported in https://users.rust-lang.org/t/giving-names-to-callback-parameters/12877.

```rust
pub fn f(callback: fn(len: usize)) {}
```

Rustdoc shows this as:

```rust
pub fn f(callback: fn(_: usize))
```

This is with rustc 1.22.0-nightly (539f2083d 2017-09-13).

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #44570.

cc @QuietMisdreavus 
cc @rust-lang/dev-tools 

Considering the impact on the `hir` libs, I'll put @eddyb as reviewer.

r? @eddyb

### CODE REVIEW

File: src/librustc/hir/mod.rs [hunk 1]
```diff
@@ -1426,7 +1426,7 @@ pub enum Ty_ {
     /// A reference (`&'a T` or `&'a mut T`)
     TyRptr(Lifetime, MutTy),
     /// A bare function (e.g. `fn(usize) -> bool`)
-    TyBareFn(P<BareFnTy>),
+    TyBareFn(P<BareFnTy>, HirVec<Spanned<Name>>),
```
Comment (eddyb) [inline src/librustc/hir/mod.rs]: When we discussed it, you were supposed to put it in `BareFnTy`. That should simplify things.

Comment (GuillaumeGomez) [reply src/librustc/hir/mod.rs]: Ah damn, got tricked by [this message](https://botbot.me/mozilla/rust-internals/2017-09-26/?msg=91596420&page=5).

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/librustc/hir/mod.rs [hunk 2]
```diff
@@ -1412,6 +1412,7 @@ pub struct BareFnTy {
     pub abi: Abi,
     pub lifetimes: HirVec<LifetimeDef>,
     pub decl: P<FnDecl>,
+    pub args_name: HirVec<Spanned<Name>>,
```
Comment (eddyb) [inline src/librustc/hir/mod.rs]: `arg_names`

File: src/librustdoc/clean/mod.rs [hunk 3]
```diff
@@ -1244,12 +1244,25 @@ impl<'a> Clean<Arguments> for (&'a [P<hir::Ty>], hir::BodyId) {
     }
 }
 
-impl<'a, A: Copy> Clean<FnDecl> for (&'a hir::FnDecl, A)
+impl<'a> Clean<Arguments> for (&'a [P<hir::Ty>], Vec<String>) {
```
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: Changing this `impl` means you've done something wrong...

File: src/librustdoc/clean/mod.rs [hunk 4]
```diff
@@ -1927,7 +1940,34 @@ impl Clean<Type> for hir::Ty {
                     _ => Infer // shouldn't happen
                 }
             }
-            TyBareFn(ref barefn) => BareFunction(box barefn.clean(cx)),
```
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: `BareFnTy` has its own `Clean` impl, you should be changing that, not this - this is why I didn't want you to change the enum variant, but the struct.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/librustdoc/clean/mod.rs [hunk 5]
```diff
@@ -1970,14 +1983,15 @@ impl<'tcx> Clean<Type> for Ty<'tcx> {
...
+                let decl = (cx.tcx.hir.local_def_id(ast::CRATE_NODE_ID), sig).clean(cx);
                 BareFunction(box BareFunctionDecl {
                     unsafety: sig.unsafety(),
                     generics: Generics {
                         lifetimes: Vec::new(),
                         type_params: Vec::new(),
                         where_predicates: Vec::new()
                     },
-                    decl: (cx.tcx.hir.local_def_id(ast::CRATE_NODE_ID), sig).clean(cx),
+                    decl: decl,
```
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: Why was this moved at all?

File: src/librustdoc/clean/mod.rs [hunk 6]
```diff
@@ -2491,7 +2505,23 @@ impl Clean<BareFunctionDecl> for hir::BareFnTy {
                 type_params: Vec::new(),
                 where_predicates: Vec::new()
             },
-            decl: (&*self.decl, &[][..]).clean(cx),
```
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: The existing impls are *already* powerful enough for this (unless I wasn't careful), all you need is to replace `&[][..]` with `&self.arg_names`.

Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs]: Not exactly:

```
no method named `clean` found for type `(&rustc::hir::FnDecl, &syntax::ptr::P<[syntax::codemap::Spanned<syntax::ast::Symbol>]>)` in the current scope
    --> src/librustdoc/clean/mod.rs:2507:50
     |
2507 |             decl: (&*self.decl, &self.arg_names).clean(cx),
     |                                                  ^^^^^
     |
     = note: the method `clean` exists but the following trait bounds were not satisfied:
             `(&[syntax::ptr::P<rustc::hir::Ty>], &syntax::ptr::P<[syntax::codemap::Spanned<syntax::ast::Symbol>]>) : clean::Clean<clean::Arguments>`
     = help: items from traits can only be used if the trait is implemented and in scope
     = note: the following trait defines an item `clean`, perhaps you need to implement it:
             candidate #1: `clean::Clean`
```

Comment (eddyb) [reply src/librustdoc/clean/mod.rs]: Sounds like you need `&self.arg_names[..]` instead?

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/librustdoc/clean/mod.rs (hunk 6 above)
Comment (eddyb) [reply src/librustdoc/clean/mod.rs]: Like, look for `ForeignItemFn`, it starts off with the same information but it *doesn't* implement the cleaning itself.

**Comment by GuillaumeGomez** [comment]
Updated.
